File: workers/dotnet/Temporalio.Omes/ActivityDispatch.cs
using Google.Protobuf.WellKnownTypes;
using Temporal.Omes.KitchenSink;

namespace Temporalio.Omes;

// Maps an ExecuteActivityAction to its registered activity name and args.
// Shared by the workflow-scheduled path and the standalone-activity path.
public static class ActivityDispatch
{
    public static (string, IReadOnlyCollection<object>) NameAndArgs(ExecuteActivityAction act)
    {
        if (act.Delay is { } delay)
        {
            return ("delay", new object[] { delay });
        }
        if (act.Payload is { } payload)
        {
            var inputData = new byte[payload.BytesToReceive];
            for (int i = 0; i < inputData.Length; i++)
            {
                inputData[i] = (byte)(i % 256);
            }
            return ("payload", new object[] { inputData, payload.BytesToReturn });
        }
        if (act.Client is { } client)
        {
            return ("client", new object[] { client });
        }
        if (act.RetryableError is { } retryableError)
        {
            return ("retryable_error", new object[] { retryableError });
        }
        if (act.Timeout is { } timeout)
        {
            return ("timeout", new object[] { timeout });
        }
        if (act.Heartbeat is { } heartbeat)
        {
            return ("heartbeat", new object[] { heartbeat });
        }
        return ("noop", Array.Empty<object>());
    }

    public static Temporalio.Common.RetryPolicy RetryPolicyFromProto(
        Temporalio.Api.Common.V1.RetryPolicy proto)
    {
        return new()
        {
            InitialInterval = proto.InitialInterval.ToTimeSpan(),
            BackoffCoefficient = (float)proto.BackoffCoefficient,
            MaximumInterval = proto.MaximumInterval?.ToTimeSpan(),
            MaximumAttempts = proto.MaximumAttempts,
            NonRetryableErrorTypes = proto.NonRetryableErrorTypes.Count == 0
                ? null
                : proto.NonRetryableErrorTypes
        };
    }
}

File: workers/ruby/workerlib/kitchensink/activity_dispatch.rb
# frozen_string_literal: true

require 'temporalio/retry_policy'

# ActivityDispatch maps an ExecuteActivityAction to the registered activity name
# and its args, and converts its retry policy and timeouts. Shared by the
# workflow-scheduled path (KitchenSinkWorkflow#launch_activity) and the
# standalone-activity client path so the two stay in sync.
module ActivityDispatch
  module_function

  # Returns [activity_name, args] for the given ExecuteActivityAction.
  # Unrecognized variants fall back to the noop activity.
  def name_and_args(exec_activity)
    case exec_activity.activity_type
    when :delay
      ['delay', [exec_activity.delay]]
    when :payload
      input_data = Array.new(exec_activity.payload.bytes_to_receive) { |i| i % 256 }.pack('C*')
      ['payload', [input_data, exec_activity.payload.bytes_to_return]]
    when :client
      ['client', [exec_activity.client]]
    when :retryable_error
      ['retryable_error', [exec_activity.retryable_error]]
    when :timeout
      ['timeout', [exec_activity.timeout]]
    when :heartbeat
      ['heartbeat', [exec_activity.heartbeat]]
    else
      ['noop', []]
    end
  end

  # Converts a proto RetryPolicy into a Temporalio::RetryPolicy, or nil if unset.
  def retry_policy_from_proto(retry_policy)
    return nil if retry_policy.nil?

    Temporalio::RetryPolicy.new(
      max_attempts: retry_policy.maximum_attempts,
      initial_interval: retry_policy.initial_interval&.then do |d|
        d.seconds + (d.nanos / 1_000_000_000.0)
      end
    )
  end

  # Converts the named protobuf Duration field to seconds, or nil if unset/zero.
  def duration_or_nil(activity_action, field)
    val = activity_action.send(field)
    return nil if val.nil? || (val.seconds.zero? && val.nanos.zero?)

    val.seconds + (val.nanos / 1_000_000_000.0)
  end
end
